src/main/java/uk/gov/hmcts/reform/cwrdapi/servicebus/TopicPublisher.java

```diff
@@ -1,62 +1,115 @@
-        log.info("{}:: Publishing message to service bus topic:: Job Id is: {}", loggingComponentName,
-                validationService.getAuditJobId());
-        if (message instanceof PublishCaseWorkerData) {
+    //The Azure library has retry mechanism inbuilt, which will try to send message until timeout.
+    // There is no need for explicit retry mechanism.
+    public void sendMessage(@NotNull PublishCaseWorkerData caseWorkerIds) {
+        ServiceBusSenderClient serviceBusSenderClient = null;
+        ServiceBusTransactionContext transactionContext = null;
+
+        try {
+            log.info("{}:: Publishing message to service bus topic:: Job Id is: {}", loggingComponentName,
+                    validationService.getAuditJobId());
+
             log.info("{}:: Job Id is: {}, Count of User Ids is: {} ",
-                    loggingComponentName,
-                    validationService.getAuditJobId(),
-                    ((PublishCaseWorkerData) message).getUserIds() != null
-                            ? ((PublishCaseWorkerData) message).getUserIds().size() : null);
-        }
+                    loggingComponentName, validationService.getAuditJobId(), caseWorkerIds.getUserIds().size());
 
-        jmsTemplate.convertAndSend(destination, message);
+            serviceBusSenderClient = messagingConfig.getServiceBusSenderClient();
+            transactionContext = serviceBusSenderClient.createTransaction();
+            publishMessageToTopic(caseWorkerIds, serviceBusSenderClient, transactionContext);
+        } catch (Exception exception) {
+            log.error("{}:: Publishing message to service bus topic failed with exception: {}:: Job Id {}",
+                    loggingComponentName, exception, validationService.getAuditJobId());
+            if (Objects.nonNull(serviceBusSenderClient) && Objects.nonNull(transactionContext)) {
+                serviceBusSenderClient.rollbackTransaction(transactionContext);
+            }
+            throw new CaseworkerMessageFailedException(CaseWorkerConstants.ASB_PUBLISH_ERROR);
+        }
+        serviceBusSenderClient.commitTransaction(transactionContext);
         log.info("{}:: Message published to service bus topic:: Job Id is: {}", loggingComponentName,
                 validationService.getAuditJobId());
     }
 
-    @Recover
-    public void recoverMessage(Exception ex) {
-        log.error("{}:: Publishing message to service bus topic failed with exception: {}:: Job Id {}",
-                loggingComponentName, ex, validationService.getAuditJobId());
-        throw new CaseworkerMessageFailedException(CaseWorkerConstants.ASB_PUBLISH_ERROR);
+    private void publishMessageToTopic(PublishCaseWorkerData caseWorkerData,
+                                      ServiceBusSenderClient serviceBusSenderClient,
+                                      ServiceBusTransactionContext transactionContext) {
+        log.info("{}:: Started publishing to topic::", loggingComponentName);
```

Comment:
Can we add job id as well for tracking purpose ?

Reply:
Okay, let me add that. It's already being added in previous step also.